track `if typing.TYPE_CHECKING` to warn about non runtime bindings

terencehonles opened the pull request:
When importing or defining values in `if typing.TYPE_CHECKING` blocks the bound names will not be available at runtime and may cause errors when used in the following way:

```python
import typing

if typing.TYPE_CHECKING:
    from module import Type  # some slow import or circular reference

def method(value) -> Type:  # the import is needed by the type checker
    assert isinstance(value, Type)  # this is a runtime error
```

This change allows pyflakes to track what names are bound for runtime use, and allows it to warn when a non runtime name is used in a runtime context.

sigmavirus24 commented on tox.ini:
```diff
@@ -8,7 +8,7 @@ deps = flake8==3.6.0
 commands =
     python --version --version
-    python -m unittest discover pyflakes
+    python -m unittest discover pyflakes {posargs}
```
This seems orthogonal

terencehonles replied:
I moved this to https://github.com/PyCQA/pyflakes/pull/624

asottile commented:
#530 predates this and has similar goals / problems -- perhaps collaborate with that PR?

terencehonles commented:
> #530 predates this and has similar goals / problems -- perhaps collaborate with that PR?

@asottile It looks like I actually ended up writing about the same code :sweat_smile: however there are a few things I have which the other one doesn't:

1. The logic handles the non runtime bindings in either the body or the else of the if/else expression
2. The runtime attribute defaults to True (it generally will be) which needs less changes in other parts of the codebase.
3. The line https://github.com/PyCQA/pyflakes/pull/530/files#diff-5df44d79406311387e0056faae7a8092d1ce9cc85c7d15420186e45a0c712283R1178-R1181 continues looking up the stack for a runtime value, but it's more likely that the non runtime value will be declared at a higher scope and therefore the code I wrote warns and aborts immediately.

It also looks like the code was waiting review for almost a year. Is there something you wanted changed from that PR? I'm willing to work with @PetterS, but I didn't realize there was an PR _that old_ for this already opened.

asottile commented:
ah shoot, that's probably on me -- I might've missed the last round of reviews on there and then it lapsed (and now conflicts)

PetterS commented:
I can revive that PR if we agree that it it likely to be merged.

terencehonles commented:
> I can revive that PR if we agree that it it is likely to be merged.
> 
> This is a problem that has been observed for real code, so would be good to fix.

It should be pretty easy to xref this PR in order to see what might also need to change since I based my PR off the most recent release (I was not aware of your PR when starting). We're 99% inline with each other's implementations.

PetterS commented:
> ah shoot, that's probably on me -- I might've missed the last round of reviews on there and then it lapsed (and now conflicts)

Merged it with master again now.

terencehonles commented:
> #530 predates this and has similar goals / problems -- perhaps collaborate with that PR?

@asottile since the PR in comment above is not going to be updated by its author

https://github.com/PyCQA/pyflakes/pull/530#issuecomment-1427121941

any chance you'd be willing to review this PR?

As mentioned above https://github.com/PyCQA/pyflakes/pull/622#issuecomment-805307482 it had used negated logic compared to the other PR in order to minimize the number of changes to the code base, and I've continued to both use and update this PR for changes on the development branch.

I can seek a 2nd approver if we can move forward with this PR.

asottile commented on pyflakes/checker.py:
```diff
@@ -1805,7 +1818,39 @@ def DICT(self, node):
+        # normalize body and orelse to a list
+        body, orelse = (
+            i if isinstance(i, list) else [i]
+            for i in (node.body, node.orelse))
```
also this is a very strange way to do this -- don't go through a generator for two assignments just do the two assignments

asottile reviewed (commented):
I don't think the tests cover all the cases you intend to demonstrate here

terencehonles commented:
Thanks for the review, I'll see when I have time to address it.

terencehonles commented on pyflakes/checker.py:
```diff
@@ -1073,12 +1078,18 @@ def handleNodeLoad(self, node, parent):
                     self.report(messages.InvalidPrintSyntax, node)
 
             try:
```
I only added a little bit here, and the exception was already broader than it needed to be. I didn't realize this was a problem, and I moved all the unrelated code (including the existing code) out of the exception to satisfy this request.

terencehonles commented on pyflakes/checker.py:
```diff
@@ -732,6 +734,7 @@ class Checker:
     nodeDepth = 0
     offset = None
     _in_annotation = AnnotationState.NONE
+    _in_type_check_guard = False
```
This is a pretty common pattern, and as you can see it already existed here. I've moved all of these into the `__init__` to match your request.

terencehonles commented:
It has taken me awhile to revisit this PR in order to address the comments (rather than just to update it), but I have addressed most of the comments.

Responding to @jayvdb's comment

> However as noted earlier in reviews, `runtime` isnt a good name. I would prefer to see a more explicit `in_type_checking` or similar, and have one name consistently used everywhere.

I was avoiding calling it `in_type_checking` since I thought there might end up being other reasons why a binding wouldn't be in the "runtime" scope and I wouldn't want to add additional flags.

> Also I suspect that this feature will be a lot cleaner `if TYPE_CHECKING` is treated as a special **scope**, which automatically means items in it are not placed into the normal scope logic. Then items in that special scope can only be considered by explicitly including them, only in the contexts when they are useful.

 I agree this might make sense, but it would require a bigger re-work and I believed I was following how `_in_annotation` was implemented.

I hope these changes are satisfactory.
